June 2018

Zhanna Zhanabekova

Jupiter Notebook 5.5.0 (Python 3.5.5, Pandas 0.23.0, scikit-learn 0.19.1)

IPython Book #1 of 3

## Objective: Predict hospital readmissions within 30 days of the last discharge among patients with diabetes. 

## Data: Diabetes 130-US Hospitals for Years 1999-2008

The dataset represents 10 years (1999-2008) of clinical care at 130 US hospitals and integrated delivery networks. It includes over 50 features representing patient and hospital outcomes. Information was extracted from the database for encounters that satisfied the following criteria.

(1)	It is an inpatient encounter (a hospital admission).

(2)	It is a diabetic encounter, that is, one during which any kind of diabetes was entered to the system as a diagnosis.

(3)	The length of stay was at least 1 day and at most 14 days.

(4)	Laboratory tests were performed during the encounter.

(5)	Medications were administered during the encounter.

The data contains such attributes as patient number, race, gender, age, admission type, time in hospital, medical specialty of admitting physician, number of lab test performed, HbA1c test result, diagnosis, number of medication, diabetic medications, number of outpatient, inpatient, and emergency visits in the year before the hospitalization, etc.

Source: http://archive.ics.uci.edu/ml/datasets/Diabetes+130-US+hospitals+for+years+1999-2008


In [28]:
# Load useful packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline


# Load data and examine basic summary statistics

In [30]:
# Load data
# directory = [[ENTER DIRECTORY PATH]]
data = pd.read_csv(directory + 'diabetic_data.csv', header = 0)
print('Number of rows (encounters) is: %s' % "{:,}".format(data.shape[0]))
print('Number of unique patients is: %s' % "{:,}".format(data.patient_nbr.nunique()))
print('Number of columns is: %s' % data.shape[1])
print("Number of numeric variables: ", data.describe(include = [np.number]).T.shape[0])
print("Number of categorical variables: ", data.describe(exclude = [np.number]).T.shape[0])


Number of rows (encounters) is: 101,766
Number of unique patients is: 71,518
Number of columns is: 50
Number of numeric variables:  13
Number of categorical variables:  37


In [31]:
# Examine numeric variables
pd.options.display.float_format = "{:,.1f}".format
data.describe(include = [np.number]).T


count          mean           std      min  \
encounter_id             101,766.0 165,201,645.6 102,640,296.0 12,522.0   
patient_nbr              101,766.0  54,330,400.7  38,696,359.3    135.0   
admission_type_id        101,766.0           2.0           1.4      1.0   
discharge_disposition_id 101,766.0           3.7           5.3      1.0   
admission_source_id      101,766.0           5.8           4.1      1.0   
time_in_hospital         101,766.0           4.4           3.0      1.0   
num_lab_procedures       101,766.0          43.1          19.7      1.0   
num_procedures           101,766.0           1.3           1.7      0.0   
num_medications          101,766.0          16.0           8.1      1.0   
number_outpatient        101,766.0           0.4           1.3      0.0   
number_emergency         101,766.0           0.2           0.9      0.0   
number_inpatient         101,766.0           0.6           1.3      0.0   
number_diagnoses         101,766.0           7.4           1.9      1.0   

                                  25%           50%           75%  \
encounter_id             84,961,194.0 152,388,987.0 230,270,887.5   
patient_nbr              23,413,221.0  45,505,143.0  87,545,949.8   
admission_type_id                 1.0           1.0           3.0   
discharge_disposition_id          1.0           1.0           4.0   
admission_source_id               1.0           7.0           7.0   
time_in_hospital                  2.0           4.0           6.0   
num_lab_procedures               31.0          44.0          57.0   
num_procedures                    0.0           1.0           2.0   
num_medications                  10.0          15.0          20.0   
number_outpatient                 0.0           0.0           0.0   
number_emergency                  0.0           0.0           0.0   
number_inpatient                  0.0           0.0           1.0   
number_diagnoses                  6.0           8.0           9.0   

                                   max  
encounter_id             443,867,222.0  
patient_nbr              189,502,619.0  
admission_type_id                  8.0  
discharge_disposition_id          28.0  
admission_source_id               25.0  
time_in_hospital                  14.0  
num_lab_procedures               132.0  
num_procedures                     6.0  
num_medications                   81.0  
number_outpatient                 42.0  
number_emergency                  76.0  
number_inpatient                  21.0  
number_diagnoses                  16.0

In [32]:
# Examine categorical variables
data.describe(exclude = [np.number]).T


count unique        top    freq
race                      101766      6  Caucasian   76099
gender                    101766      3     Female   54708
age                       101766     10    [70-80)   26068
weight                    101766     10          ?   98569
payer_code                101766     18          ?   40256
medical_specialty         101766     73          ?   49949
diag_1                    101766    717        428    6862
diag_2                    101766    749        276    6752
diag_3                    101766    790        250   11555
max_glu_serum             101766      4       None   96420
A1Cresult                 101766      4       None   84748
metformin                 101766      4         No   81778
repaglinide               101766      4         No  100227
nateglinide               101766      4         No  101063
chlorpropamide            101766      4         No  101680
glimepiride               101766      4         No   96575
acetohexamide             101766      2         No  101765
glipizide                 101766      4         No   89080
glyburide                 101766      4         No   91116
tolbutamide               101766      2         No  101743
pioglitazone              101766      4         No   94438
rosiglitazone             101766      4         No   95401
acarbose                  101766      4         No  101458
miglitol                  101766      4         No  101728
troglitazone              101766      2         No  101763
tolazamide                101766      3         No  101727
examide                   101766      1         No  101766
citoglipton               101766      1         No  101766
insulin                   101766      4         No   47383
glyburide-metformin       101766      4         No  101060
glipizide-metformin       101766      2         No  101753
glimepiride-pioglitazone  101766      2         No  101765
metformin-rosiglitazone   101766      2         No  101764
metformin-pioglitazone    101766      2         No  101765
change                    101766      2         No   54755
diabetesMed               101766      2        Yes   78363
readmitted                101766      3         NO   54864

In [33]:
# Check if encounter_id is unique. It is (zero duplicates).
data[data.duplicated(['encounter_id'], keep=False)].shape[0]


0

In [34]:
# Confirm that patient_nbr is not unique (i.e., some patients have multiple encounters)
data[data.duplicated(['patient_nbr'], keep=False)].shape[0]  # Observe a non-zero value


47021

# Create target variable

The task is to "predict which encounters will be followed by the patient being readmitted to the hospital within 30 days."

Note that B. Strack et al (the team that created and donated the dataset to the UCI Machine Learning Depository) "defined the readmission attribute (outcome) as having two values: “readmitted,” if the patient was readmitted within 30 days of discharge or “otherwise,” which covers both readmission after 30 days and no readmission at all." 

Next, I examine the 'readmitted' variable and construct the target / outcome variable.

In [35]:
# Tabulate the readmitted variable (using counts)
outcome_tabulated = pd.crosstab(index = data.readmitted, columns = 'count')
outcome_tabulated


col_0       count
readmitted       
<30         11357
>30         35545
NO          54864

In [36]:
# Tabulate the readmitted variable (using %)
pd.options.display.float_format = "{:,.2f}".format
outcome_tabulated/outcome_tabulated.sum()


col_0       count
readmitted       
<30          0.11
>30          0.35
NO           0.54

Note that about 11% of all encounters in the raw data result in readmissions within 30 days.

In [37]:
# Create the binary target / outcome variable: 1 if the encounter results in an readmission within 30 days; 0 otherwise
target = pd.get_dummies(data.readmitted)
data = pd.concat([data, target], axis=1)

# Drop dummies created for categories '>30' and 'NO'
data = data.drop(['>30', 'NO'], axis=1) 

# Rename the target variable (currently named as '<30') as target and convert it to integer (from float)
data = data.rename(columns= {'<30': 'target'})
data.target = data.target.apply(np.int64)

# Drop encounters that resulted in death or discharge to hospice care

Next, I follow B. Strack et al and exclude encounters that resulted in death or discharge to hospice care. Keeping these observations would mean that our analysis includes encounters that can never result in a readmission and thus would inflate the possibility of the negative outcome (no readmission or readmission after 30 days). 

B. Strack et al "removed all encounters that resulted in either discharge to a hospice or patient death, to avoid biasing [their] analysis."




After examining categories of the discharge_disposition_id variable, I decide to drop the following categories:

11 - Expired

13 - Hospice / home

14 - Hospice / medical facility

19 - Expired at home. Medicaid only, hospice

20 - Expired in a medical facility. Medicaid only, hospice

21 - Expired, place unknown. Medicaid only, hospice.

In [38]:
# Number of unique patients before I drop encounters with disposition to hospice care or death
data.patient_nbr.nunique()


71518

In [39]:
# Remove the relevant categories
for i in [11,13,14,19,20,21]:
    data = data[data.discharge_disposition_id !=i]
    

In [40]:
# Check the number of unique patients post-removal
data.patient_nbr.nunique()


69990

After removing the hospice and death categories, the number of unique patients drops to 69,990 (from 71,518). Note that the clean dataset of B. Stack et al has 69,984 unique patients, so this this is pretty close (a difference of only 6 patients).

# Keeping only one encounter per patient

Next comes the interesting part of deduping the dataset, so that it only contains one encounter per patient.

According to B. Strack et al: "The preliminary dataset contained multiple inpatient visits for some patients and the observations could not be considered as statistically independent, an assumption of the logistic regression model. We thus used only one encounter per patient; in particular, we considered only the first encounter for each patient as the primary admission and determined whether or not they were readmitted within 30 days."


We also have a tip that "[a] patient can appear in the dataset multiple times!  And it is not necessarily chronological with "encounter_id" (e.g. a smaller "encounter_id" does not necessarily represent an earlier admission)."

So the task of keeping only 1 encounter per patient is not very straightforward since we do not know the date of admission, and hence it is not possible to figure out the first encounter for each patient.

For the purpose of this exercise only, I propose the following ad-hoc solution:

Step 1:
- Keep all patients with only 1 encounter as is (note that some of these are coded as being readmitted within 30 days)
- For patients with only 1 readmission within 30 days, select that one encounter with target=1
- Identify all patients with multiple encounters and multiple readmissions within 30 days, and keep only those observations with target=1 (i.e., here we still have more than 1 observation per patient with more than one target=1 rows)
- Identify all patients with more than 1 encounter but zero readmissions within 30 days (here we also have multiple observations per patient, all with target=0)

Step 2:
- Then keep only those observations with the smallest number of missing values across all columns

Step 3:
- At this point, we still have some patients that show up more than once. For the purpose of this exercise, after sorting data on patient_nbr, I select and keep the first observation for each patient (this affects only patients with multiple observations).


In [41]:
# For each patient, calculate the total number of readmissions within 30 days and the total number of encounters
target_groupped = data.groupby(data.patient_nbr).agg({"target":['sum', 'count']})
target_groupped.columns = target_groupped.columns.droplevel(level=0) # drop the top row (redundant)
target_groupped = target_groupped.rename(columns=
                       {"sum": "readmissions_per_patient", "count": "encounters_per_patient"})  # rename new aggregate vars
target_groupped.readmissions_per_patient = target_groupped.readmissions_per_patient.apply(np.int64) # set as integer (vs float)


In [42]:
# Add the new aggregate variables ('readmissions_per_patient','encounters_per_patient') to the main dataset. Call it 'result'
result = data.join(target_groupped, on='patient_nbr')


In [44]:
# Sort results by patient_nbr
result = result.sort_values(by=['patient_nbr'])


In [45]:
# Tabulate the number of readmissions within 30 days (counts) (per patient) [deduped data]
result_unique_patients = result.drop_duplicates(['patient_nbr'], keep='first')
readmiss = pd.crosstab(index = result_unique_patients.readmissions_per_patient, columns = 'count')
readmiss


col_0                     count
readmissions_per_patient       
0                         61185
1                          7277
2                          1041
3                           290
4                            94
5                            41
6                            24
7                            14
8                             8
9                             2
10                            3
11                            3
12                            6
22                            1
23                            1

In [46]:
# Tabulate the number of readmissions within 30 days (%) (per patient) [deduped data]
readmiss/readmiss.sum()


col_0                     count
readmissions_per_patient       
0                          0.87
1                          0.10
2                          0.01
3                          0.00
4                          0.00
5                          0.00
6                          0.00
7                          0.00
8                          0.00
9                          0.00
10                         0.00
11                         0.00
12                         0.00
22                         0.00
23                         0.00

In [47]:
# Tabulate the number of encounters per patient (%) [deduped data]
encounters= pd.crosstab(index = result_unique_patients.encounters_per_patient, columns = 'count')
encounters/encounters.sum()


col_0                   count
encounters_per_patient       
1                        0.77
2                        0.15
3                        0.05
4                        0.02
5                        0.01
6                        0.00
7                        0.00
8                        0.00
9                        0.00
10                       0.00
11                       0.00
12                       0.00
13                       0.00
14                       0.00
15                       0.00
16                       0.00
17                       0.00
18                       0.00
19                       0.00
20                       0.00
21                       0.00
22                       0.00
23                       0.00
28                       0.00
40                       0.00

Observations: 
- About 23% of all patients have more than one encounter (inpatient visits). 
- About 13% of all patients were readmitted within 30 days.
- There are some patients that have more than 3 encounters and/or readmissions within 30 days.

In [48]:
# Tabulate the number of encounters per patient against the number of readmissions per patient on NON-deduped dataset. 
# This table is helpful in explaining my deduping process
pd.crosstab(index = result.encounters_per_patient, columns = result.readmissions_per_patient)


readmissions_per_patient     0     1    2    3    4   5   6   7   8   9   10  \
encounters_per_patient                                                         
1                         51362  2287    0    0    0   0   0   0   0   0   0   
2                         14568  5520  348    0    0   0   0   0   0   0   0   
3                          5136  3570  903   90    0   0   0   0   0   0   0   
4                          2128  2224  796  232   36   0   0   0   0   0   0   
5                           915  1360  795  300   80   5   0   0   0   0   0   
6                           390   696  582  270   78  12   0   0   0   0   0   
7                           154   392  399  231  112  49   7   0   0   0   0   
8                           152   176  176  200  112  56  24   0   0   0   0   
9                            45    81  189  180   54   0  45   9   0   0   0   
10                            0    50   60   90   60  70  40  20  20   0   0   
11                            0    33   11   33   44  44  33  33  11   0   0   
12                           12    12   24   24   48  72  24   0   0   0   0   
13                            0     0    0   26   39  26  26  13  13   0   0   
14                            0     0   14   14   28   0  14   0   0   0   0   
15                            0     0   15   15   15  30  15  15  15  15   0   
16                            0     0    0    0    0  32  16   0   0  16   0   
17                            0     0    0    0    0  17   0  17   0   0  17   
18                            0     0    0   18    0   0   0   0  36   0  36   
19                            0     0    0    0    0   0  19   0  19   0   0   
20                            0     0    0    0    0   0   0  80   0   0   0   
21                            0     0    0    0    0   0   0   0   0   0   0   
22                            0     0    0    0    0   0   0   0   0   0   0   
23                            0     0    0    0    0   0   0  23   0   0   0   
28                            0     0    0    0    0   0   0   0   0   0   0   
40                            0     0    0    0    0   0   0   0   0   0   0   

readmissions_per_patient  11  12  22  23  
encounters_per_patient                    
1                          0   0   0   0  
2                          0   0   0   0  
3                          0   0   0   0  
4                          0   0   0   0  
5                          0   0   0   0  
6                          0   0   0   0  
7                          0   0   0   0  
8                          0   0   0   0  
9                          0   0   0   0  
10                         0   0   0   0  
11                         0   0   0   0  
12                         0   0   0   0  
13                         0   0   0   0  
14                         0   0   0   0  
15                         0   0   0   0  
16                         0   0   0   0  
17                         0   0   0   0  
18                        18  18   0   0  
19                         0   0   0   0  
20                        20  20   0   0  
21                         0  21   0   0  
22                         0  44   0   0  
23                        23  23   0   0  
28                         0   0  28   0  
40                         0   0   0  40

In [49]:
# Step 1. Create the 'keep' variable to identify which encounters should be kept in the final deduped dataset. Specifically,

# keep1
### Keep all patients with only 1 encounter as is; and
### For patients with only 1 readmission within 30 days, select that one encounter with target=1
result['keep1']= np.where((result.encounters_per_patient == 1) | \
                         ((result.readmissions_per_patient == 1) & (result.target==1)), 1, 0)

# keep2
### For patients with more than 1 encounter and more than 1 readmission, keep all those observations with target=1 (for now) 
result['keep2']= np.where((result.encounters_per_patient > 1) & (result.readmissions_per_patient >1) & (result.target==1), 1, 0)

# keep3
### Identify all patients with more than 1 encounter but zero readmissions (for now)
result['keep3']= np.where((result.encounters_per_patient > 1) & (result.readmissions_per_patient ==0), 1, 0)

# Var keep is the sum of keep1, keep2, keep3 (i.e., identifies all types of observations described above)
result['keep'] = result.keep1 + result.keep2 + result.keep3

# Tabulate variable keep
pd.crosstab(index = result.keep, columns = 'count')

col_0  count
keep        
0      13167
1      86176

In [50]:
# Keep only the observations identified by keep = 1
result = result[result.keep ==1]


In [ ]:
# Step 2. Then keep only those observations with the smallest number of missing values across all columns

In [51]:
# Identify the number of missing values in each row
result['no_of_missing'] = result.isin({'?'}).sum(1)

# Tabulate the result
print(result.no_of_missing.value_counts())

# For each patient, identify rows with the smallest number of missing values
min_missing = result.groupby(result.patient_nbr).agg({"no_of_missing":'min'})
min_missing = min_missing.rename(columns={"no_of_missing": "min_missing"})
result = result.join(min_missing, on='patient_nbr')

# Keep only those rows with the minimum number of missing values (this will only affect patients with multiple obs)
result = result[result.min_missing == result.no_of_missing]
print(result.shape[0])

# At this point, we still have some patients that show up more than once. For the purpose of this exercise, after (previously) 
# sorting data on patient_nbr, I select and keep the first observation for each patient (this affects only patients 
# with multiple obs)
result= result.drop_duplicates(['patient_nbr'], keep='first')
print("Number of observations: ", "{:,}".format(result.shape[0]))
print("Number of patients: ", "{:,}".format(result.patient_nbr.nunique()))

# By this point, we have exactly one encounter per patient (n=69,900)


2    45949
1    23129
3    15537
0      997
4      537
5       27
Name: no_of_missing, dtype: int64
82837
Number of observations:  69,990
Number of patients:  69,990


In [52]:
# Check columns
result.columns


Index(['encounter_id', 'patient_nbr', 'race', 'gender', 'age', 'weight',
       'admission_type_id', 'discharge_disposition_id', 'admission_source_id',
       'time_in_hospital', 'payer_code', 'medical_specialty',
       'num_lab_procedures', 'num_procedures', 'num_medications',
       'number_outpatient', 'number_emergency', 'number_inpatient', 'diag_1',
       'diag_2', 'diag_3', 'number_diagnoses', 'max_glu_serum', 'A1Cresult',
       'metformin', 'repaglinide', 'nateglinide', 'chlorpropamide',
       'glimepiride', 'acetohexamide', 'glipizide', 'glyburide', 'tolbutamide',
       'pioglitazone', 'rosiglitazone', 'acarbose', 'miglitol', 'troglitazone',
       'tolazamide', 'examide', 'citoglipton', 'insulin',
       'glyburide-metformin', 'glipizide-metformin',
       'glimepiride-pioglitazone', 'metformin-rosiglitazone',
       'metformin-pioglitazone', 'change', 'diabetesMed', 'readmitted',
       'target', 'readmissions_per_patient', 'encounters_per_patient', 'keep1',
       'keep

In [53]:
# Drop some redundant columns
result = result.drop(['readmissions_per_patient', 'encounters_per_patient', 'keep1',
       'keep2', 'keep3', 'keep', 'no_of_missing', 'min_missing'], axis=1)


In [141]:
# Save the 'result' dataset and continue working in the next ipython notebook, where I do more data cleaning and processing
# The pickle will be saved in the same folder where the IPython Notebook is located
import pickle
with open('result.pickle', 'wb') as handle:
    pickle.dump(result, handle, protocol=pickle.HIGHEST_PROTOCOL)
    
# to open, use this command:
# result = pd.read_pickle("result_book2.pickle" )

Continue to IPython Notebook #2 of 3